builtin/v10/check.go

```diff
@@ -0,0 +1,359 @@
+				return err
+			}
+			summary, msgs := power.CheckStateInvariants(&st, tree.Store)
+			acc.WithPrefix("power: ").AddAll(msgs)
+			powerSummary = summary
+		case actorCodes[manifest.MinerKey]:
+			var st miner.State
+			if err := tree.Store.Get(tree.Store.Context(), actor.Head, &st); err != nil {
+				return err
+			}
+			summary, msgs := miner.CheckStateInvariants(&st, tree.Store, actor.Balance)
+			acc.WithPrefix("miner: ").AddAll(msgs)
+			minerSummaries[key] = summary
+		case actorCodes[manifest.MarketKey]:
+			var st market.State
+			if err := tree.Store.Get(tree.Store.Context(), actor.Head, &st); err != nil {
+				return err
+			}
+			summary, msgs := market.CheckStateInvariants(&st, tree.Store, actor.Balance, priorEpoch)
+			acc.WithPrefix("market: ").AddAll(msgs)
+			marketSummary = summary
+		case actorCodes[manifest.PaychKey]:
+			var st paych.State
+			if err := tree.Store.Get(tree.Store.Context(), actor.Head, &st); err != nil {
+				return err
+			}
+			summary, msgs := paych.CheckStateInvariants(&st, tree.Store, actor.Balance)
+			acc.WithPrefix("paych: ").AddAll(msgs)
+			paychSummaries = append(paychSummaries, summary)
+		case actorCodes[manifest.MultisigKey]:
+			var st multisig.State
+			if err := tree.Store.Get(tree.Store.Context(), actor.Head, &st); err != nil {
+				return err
+			}
+			summary, msgs := multisig.CheckStateInvariants(&st, tree.Store)
+			acc.WithPrefix("multisig: ").AddAll(msgs)
+			multisigSummaries = append(multisigSummaries, summary)
+		case actorCodes[manifest.RewardKey]:
+			var st reward.State
+			if err := tree.Store.Get(tree.Store.Context(), actor.Head, &st); err != nil {
+				return err
+			}
+			summary, msgs := reward.CheckStateInvariants(&st, tree.Store, priorEpoch, actor.Balance)
+			acc.WithPrefix("reward: ").AddAll(msgs)
+			rewardSummary = summary
+		case actorCodes[manifest.VerifregKey]:
+			var st verifreg.State
+			if err := tree.Store.Get(tree.Store.Context(), actor.Head, &st); err != nil {
+				return err
+			}
+			summary, msgs := verifreg.CheckStateInvariants(&st, tree.Store, priorEpoch)
+			acc.WithPrefix("verifreg: ").AddAll(msgs)
+			verifregSummary = summary
+		case actorCodes[manifest.DatacapKey]:
+			var st datacap.State
+			if err := tree.Store.Get(tree.Store.Context(), actor.Head, &st); err != nil {
+				return err
+			}
+			summary, msgs := datacap.CheckStateInvariants(&st, tree.Store)
+			acc.WithPrefix("datacap: ").AddAll(msgs)
+			datacapSummary = summary
+		case actorCodes[manifest.EvmKey]:
+			var st evm.State
+			if err := tree.Store.Get(tree.Store.Context(), actor.Head, &st); err != nil {
+				fmt.Println("evm invariant error = ", err)
+				return err
+			}
+			msgs := evm.CheckStateInvariants(&st, tree.Store)
+			acc.WithPrefix("evm: ").AddAll(msgs)
+		case actorCodes[manifest.PlaceholderKey]:
+			if err := tree.Store.Get(tree.Store.Context(), actor.Head, &builtin.EmptyState{}); err != nil {
```

Comment:
I think the correct assertion to make for these three actor types is that the head is always equal to the "empty object" CID. You can see how it's constructed [here](https://github.com/filecoin-project/go-state-types/blob/890a3642f2cbd8fee141f682d17f74db1e8abf7e/builtin/v10/migration/top.go#L354), but I would propose making this a constant in `shared.go` replacing the `EmptyState{}` struct, and referencing that.

Reply:
done